Implement pushd, popd, dirs

wrywerytwreywery opened the pull request:
This changeset implements `pushd`, `popd`, and `dirs`.  This implementation is similar to the bash version, but with the meaning of `+` and `-` switched (as though `$PUSHD_MINUS` were set), since I think that's how most people use it anyway.  I think I got most of the features right, but I would be glad for any feedback.

This does not implement all of the features suggested in #48.  In particular, those involving changes to `cd` and tab completion aren't included here, but will come later.

scopatz commented on xonsh/built_ins.py:
```diff
@@ -414,7 +414,7 @@ def load_builtins(execer=None):
     builtins.execx = None if execer is None else execer.exec
     builtins.compilex = None if execer is None else execer.compile
-    builtins.aliases = Aliases(DEFAULT_ALIASES)
+    builtins.default_aliases = builtins.aliases = Aliases(DEFAULT_ALIASES)
```
`default_aliases` also needs to be removed in the `unload_builtins()` function.

scopatz commented on xonsh/dirstack.py:
```diff
@@ -0,0 +1,216 @@
+
+    try:
+        args = pushd_parser.parse_args(args)
+    except SystemExit:
```
Does this actually happen?

scopatz replied:
I am curious :)

wrywerytwreywery replied:
Yes it does, actually!  If there is an error when parsing, `ArgumentParser` prints the help message and then exits using `SystemExit`.

scopatz replied:
That is so silly!

scopatz commented on xonsh/dirstack.py:
```diff
@@ -0,0 +1,216 @@
+DIRSTACK = []
+
+
+def pushd(args, stdin=None):
```
docstring needed.

scopatz commented on xonsh/dirstack.py:
```diff
@@ -0,0 +1,216 @@
+    return None, None
+
+
+def popd(args, stdin=None):
```
docstring needed.

scopatz commented on xonsh/dirstack.py:
```diff
@@ -0,0 +1,216 @@
+    return None, None
+
+
+def dirs(args, stdin=None):
```
docstring needed.

scopatz commented on xonsh/dirstack.py:
```diff
@@ -0,0 +1,216 @@
+        try:
+            num = int(args.dir[1:])
+            assert num >= 0
+        except:
```
It is much better to do `if num >= 0` than an except.  Also, I think that int will only ever throw a value error.  So that error is what should be caught.

wrywerytwreywery replied:
Hmmm...the reason I kept them together there and used the catch-all was that I wanted the same error message, both in the case that a negative number is supplied, and in the case that what the user put in isn't a number at all.  To me, this seemed cleaner than repeating the error code in a second place...

scopatz replied:
Well, I don't think we should really ever be using catch-all exceptions.  Maybe that is a little puritanical :) And in general, I think if we can test a condition rather than using an exception, we should. Exceptions being for exceptional cases, etc.

scopatz commented:
Thanks for implementing this @wrywerytwreywery! There are some issues I have commented on inline. Also, a file in the docs that hooks this new module up through autodocs is needed in addtion to what is here.

scopatz commented:
Probably should document the PUSHD_MINUS like behaviour somewhere too.

wrywerytwreywery commented:
Okay, I just pushed a whole bunch of things:
* fixed a bug in `pushd`'s handling of the `-n` flag
* unloaded `default_aliases` from `builtins` (I think)
* switched back to the conventional meanings of `+` and `-`
* added support for `$PUSHD_MINUS` to switch the meanings of `+` and `-`
* added support for `$DIRSTACK_SIZE` to set a maximum number of directories to remember (default 20)
* added support for `$PUSHD_SILENT` to prevent `pushd` and `popd` from calling `dirs` after they are done
* added module to api documentation, and added docstrings for the functions
* refactored per your suggestions

I think this will be ready for a look once travis finishes.  

It might also be worth adding an explanation to the docs.  Is the tutorial the right place for this?

scopatz commented:
> It might also be worth adding an explanation to the docs. Is the tutorial the right place for this?

I think we are now at the point where the aliases deserve their own page.

scopatz commented on xonsh/built_ins.py:
```diff
@@ -437,6 +437,7 @@ def unload_builtins():
              '__xonsh_pyexit__', '__xonsh_pyquit__', 
              '__xonsh_subproc_captured__', '__xonsh_subproc_uncaptured__', 
              '__xonsh_execer__', 'evalx', 'execx', 'compilex', 
+             'default_aliases'
```
yep, this is correct :)

scopatz commented:
Thanks a ton @wrywerytwreywery!
